### PR TITLE
Consolidate webdav code - move all to one app

### PR DESCRIPTION
**Pull request by DeepDiver1975**
Starting point to get more control over the WebDAV code.

@PVince81 @icewind1991 @MorrisJobke 

Due to renaming of the webdav classes this has an impact on contacts and calendar.

-> calendar and contacts apps require to be patched

### CODE REVIEW

**Comment by MorrisJobke** [comment]
Did some testing and the webdav endpoint still behaves as before. Also the code looks goot :+1:

File: apps/dav/appinfo/info.xml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0"?>
+<info>
+	<id>dav</id>
+	<name>WebDAV</name>
+	<description>ownCloud WebDAV endppint</description>
```
Comment (icewind1991) [inline apps/dav/appinfo/info.xml]: typo

**Comment by icewind1991** [comment]
Code looks good

**Comment by PVince81** [comment]
@DeepDiver1975 9.0 ?

**Comment by DeepDiver1975** [comment]
I prefer 8.2 - mainly just because .... :Wink:

**Comment by karlitschek** [comment]
if it is 'ready' why not 8.2. But I can't judge at the moment.

**Comment by DeepDiver1975** [comment]
autoload fooo

````
https://ci.owncloud.org/job/server-master-linux-php7-ci/database=sqlite,label=SLAVE/5722/console

**Comment by DeepDiver1975** [comment]
Unit tests are passing now - please review

@PVince81 @icewind1991 @PVince81 THX

**Comment by schiesbn** [comment]
Does this mean that every app which needs webdav will depend on the dav-app? Until now I thought that we want to avoid such app dependencies... just asking.

**Comment by DeepDiver1975** [comment]
> Does this mean that every app which needs webdav will depend on the dav-app? Until now I thought that we want to avoid such app dependencies... just asking.

the dependency will be on http level
remote.php/webdav/ is the public api which exposes the webdav functionality - no matter in which app the code lives.

The interesting question at the end is: how do apps add functionalities to webdav - we have to come up with an idea soon-ish.

**Comment by phsc** [comment]
Does this mean, that I only need the webdav-app, if I just want to use CalDAV and CardDAV sync? The contacts and calendar app won't be needed anymore for this purpose?

**Comment by DeepDiver1975** [comment]
@phsc yes

File: apps/dav/appinfo/info.xml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	<licence>AGPL</licence>
+	<author>owncloud.org</author>
+	<requiremin>8.2</requiremin>
```
Comment (rullzer) [inline apps/dav/appinfo/info.xml]: 9.0 now I assume

File: apps/dav/appinfo/v1/webdav.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	\OC::$server->getUserSession(),
+	\OC::$server->getMountManager(),
+	\OC::$server->getTagManager()
+);
+
+// Backends
+$authBackend = new \OCA\DAV\Connector\Sabre\Auth();
+$requestUri = \OC::$server->getRequest()->getRequestUri();
+
+$server = $serverFactory->createServer($baseuri, $requestUri, $authBackend, function() {
```
Comment (rullzer) [inline apps/dav/appinfo/v1/webdav.php]: `$baseuri` is not defined here...

Comment (icewind1991) [inline apps/dav/appinfo/v1/webdav.php]: I believe that gets defined in remote.php and is available in the scope from there

Comment (rullzer) [inline apps/dav/appinfo/v1/webdav.php]: Ah you are right... I must say I do not like that... but that is not the scope of this PR.

File: apps/dav/lib/connector/sabre/lockplugin.php [hunk 4]
```diff
@@ -21,16 +21,14 @@
 
 namespace OC\Connector\Sabre;
```
Comment (rullzer) [inline apps/dav/lib/connector/sabre/lockplugin.php]: Should be `OCA\DAV\Connector\Sabre`

**Comment by rullzer** [comment]
Webdav Access with cadaver now works for me.
On to more testing.

**Comment by rullzer** [comment]
Did some more webdav stuff and let the sync client sync against it. All seems to work as expected.

Code makes sense. :+1:

**Comment by MorrisJobke** [comment]
@rullzer @icewind1991 I fixed your both comments

**Comment by schiesbn** [comment]
> the dependency will be on http level
> remote.php/webdav/ is the public api which exposes the webdav functionality - no matter in which app the code lives.

but the API will only work if the app is enabled? Is this a good idea? I can already see people disabling the app or deleting it completely and wonder why sync no longer works or why they get strange error messages, exceptions, etc.

Yesterday we talked about this in Stuttgart a bit. It is great to have more modularity but from a user point of view a app is always something which is self-contained and doesn't influence other functionality. So we have to expect that people disable it or even delete it either by purpose or because they doesn't know exactly what it does and just think that they don't need such a app. That's how most people perceive the apps folder, a folder where they can add and remove apps as they wish without breaking other stuff. Therefore I'm not really sure if this is the right place for such a central component like the webdav implementation.

**Comment by rullzer** [comment]
@schiesbn I agree. However we can argue that the apps we ship already have a special status. Disabling the files app is also not considered a good idea. 

Maybe we should dump shipped apps in a separate dir... so the apps directory is really for "extra" apps.  And the shipped apps get a kind of special status.

**Comment by DeepDiver1975** [comment]
> but the API will only work if the app is enabled? So we have a core API which might work or might not work depending on a app state?

I fully understand that - I have been thinking about this as well.
We can fix this the same way as we did this with the files app which cannot be disabled at all.
As of today this is done with a hardcoded hack specific to the  files app.

In a follow up PR we can discuss to add a info element or a file similar to shipped.json which holds this information.

But besides this: we have to move more functionalities into apps in the next releases - core is already a big beast and we can only conquer it if we divide it!

**Comment by karlitschek** [comment]
i like modular so having this in an app is a good idea. But I also agree that a user should not be able to disable this. There are usecases to disable, encryption, version, trash and so on but this would be special.
Maybe an app that can disabled like files as @DeepDiver1975 said is the solution.

**Comment by DeepDiver1975** [comment]
> Maybe an app that can disabled like files as @DeepDiver1975 said is the solution.

I'll come up with an approach in its own PR

**Comment by DeepDiver1975** [comment]
@schiesbn @icewind1991 @PVince81 @karlitschek I need at least one aditional :+1: - THX

**Comment by PVince81** [comment]
I'm ok with this :+1:

**Comment by karlitschek** [comment]
very sweet :+1: 
So we can release 9.0 together with 8.2 it seems ;-) ;-)

**Comment by MorrisJobke** [comment]
```
{"reqId":"LwoJmEDGuIlnguWED594","remoteAddr":"::1","app":"remote","message":"Autoload path not allowed: owncloud\/master\/apps\/dav\/lib\/connector\/sabre\/serverfactory.php","level":4,"time":"2015-10-16T14:40:04+00:00"}
```

And in the browser:

```
Technische Details
Entfernte Adresse: ::1
Anfragekennung: vm2cnG3NcB3/Byqp3A9h
Typ: OCP\AutoloadNotAllowedException
Code: 0
Nachricht: Autoload path not allowed: master/apps/dav/lib/connector/sabre/serverfactory.php
Datei: master/lib/autoloader.php
Zeile: 138

Ablauf
#0 master/lib/autoloader.php(158): OC\Autoloader->isValidPath('/Users/morrisjo...')
#1 [internal function]: OC\Autoloader->load('OCA\\DAV\\Connect...')
#2 master/apps/files/appinfo/remote.php(35): spl_autoload_call('OCA\\DAV\\Connect...')
#3 master/remote.php(137): require_once('/Users/morrisjo...')
#4 {main}
```

This is the result of a PROPFIND with curl :(

**Comment by DeepDiver1975** [comment]
> master/apps/files/appinfo/remote.php(35)

hmmmm ....

**Comment by icewind1991** [comment]
make sure the app is enabled

**Comment by DeepDiver1975** [comment]
> master/apps/files/appinfo/remote.php(35)

that file needs to be removed ... pr to come

**Comment by DeepDiver1975** [comment]
> that file needs to be removed ... pr to come

https://github.com/owncloud/core/pull/19838

**Comment by DeepDiver1975** [comment]
> make sure the app is enabled

the app is enabled by default

**Comment by DeepDiver1975** [comment]
> the app is enabled by default

fix is here https://github.com/owncloud/core/pull/19843

**Pull request merged**